nifi-nar-bundles/nifi-iotdb-bundle/nifi-iotdb-processors/src/main/java/org/apache/nifi/processors/PutIoTDBRecord.java

```diff
@@ -0,0 +1,293 @@
+                    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+                    .required(false)
+                    .defaultValue("1024")
+                    .build();
+
+    private static final String ROOTPREFIX = "root.";
+
+    @Override
+    protected void init(final ProcessorInitializationContext context) {
+        super.init(context);
+        descriptors.add(RECORD_READER_FACTORY);
+        descriptors.add(SCHEMA);
+        descriptors.add(PREFIX);
+        descriptors.add(ALIGNED);
+        descriptors.add(MAX_ROW_NUMBER);
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return descriptors;
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return relationships;
+    }
+
+    @Override
+    public void onTrigger(ProcessContext processContext, ProcessSession processSession)
+            throws ProcessException {
+
+        FlowFile flowFile = processSession.get();
+
+        if (flowFile == null) {
+            return;
+        }
+
+        String schemaProperty = processContext.getProperty(SCHEMA).evaluateAttributeExpressions(flowFile).getValue();
+        String alignedProperty = processContext.getProperty(ALIGNED).evaluateAttributeExpressions(flowFile).getValue();
+        String maxRowNumberProperty = processContext.getProperty(MAX_ROW_NUMBER).evaluateAttributeExpressions(flowFile).getValue();
+        String prefix = processContext.getProperty(PREFIX).evaluateAttributeExpressions(flowFile).getValue();
+        final RecordReaderFactory recordParserFactory =
+                processContext
+                        .getProperty(RECORD_READER_FACTORY)
+                        .asControllerService(RecordReaderFactory.class);
+
+        if (!prefix.startsWith(ROOTPREFIX) || !prefix.endsWith(".")) {
+            getLogger().error("The prefix is not begin with root and end with .", flowFile);
+            processSession.transfer(flowFile, REL_FAILURE);
+        }
+
+        final boolean aligned = Boolean.valueOf(alignedProperty);
+        int maxRowNumber = Integer.valueOf(maxRowNumberProperty);
+
+        try (final InputStream inputStream = processSession.read(flowFile);
+             final RecordReader recordReader =
+                     recordParserFactory.createRecordReader(flowFile, inputStream, getLogger())) {
+            IoTDBSchema schema = getSchema(schemaProperty, recordReader);
+
+            List<String> fieldNames = schema.getFieldNames(prefix);
+
+            boolean needInitFormatter = schema.getTimeType() != IoTDBSchema.TimeType.LONG;
+
+            HashMap<String, Tablet> tablets = generateTablets(schema, prefix, maxRowNumber);
+            DateTimeFormatter format = null;
+
+            Record record;
+
+            while ((record = recordReader.nextRecord()) != null) {
+                Object[] values = record.getValues();
+                long timestamp = getTimeStamp(values[0], needInitFormatter, format);
```

Comment:
This approach still requires the timestamp as the first field in the record, instead of making it configurable. As mentioned previously, it seems like the general approach needs to be re-evaluated since this Processor currently requires very strict Record formatting, when it should be configurable.